### DISCUSSION TITLE
Contributing to paperless-ng

### DISCUSSION THREAD

**Discussion by jcpsantiago** [jonaswinkler/paperless-ng#169] [General]
👋 I've been adding some issues and comments here and there, but would like to contribute more actively. 
I'm more of a Clojure person and, although I have some experience with Python, I have never worked with Django before. 
I could give a hand in smaller issues, documentation, maybe deployment.

Taking this into account, any issues/things I could give you a hand with @jonaswinkler?

**Comment by jonaswinkler** [comment]
Well, off the top of my head, these are the main concerns I have with the project right now:

- The most pressing issue, by far, is the fact that the entire front end is untested (apart from me clicking through all the pages occasionally). But since that is still subject to change, I haven't done anything in that direction. I don't even know how front end testing works.
- I want to automate the entire build and publishing process, see #151.
- ~Parts of the front end feel really sluggish, especially the document list with 100 cards. I'm looking into ways to make that faster.~
- I'm also not entirely happy with the look of the management pages, but they are functional, and get the job done. Some filtering would probably be nice, especially on the correspondent list, which can get quite long.
- There isn't all that much to do on the django part, I've been primarily busy writing test cases for most of the components. But that's nearly done as well.
- There's still lots of code in the Django part that I want to get rid of, most notably the entire encryption stuff. That part contains a couple of issues (primarily the fact that opened file handles on encrypted files are apparently never closed), and I'm not very keen on fixing code that no one should use anyway.
- In the long run I want to make a nice interface for configuring mail accounts and rules, since I feel that's not very accessible at the moment.
- The documentation could really use some polishing here and there, especially since some users weren't able to find the section about how the automatic matching algorithm works. I've linked to that from the index, I don't know.

I've also played with the idea to make a survey on how paperless is used, to get a feel for which part I should work on more. Questions such as which features are used and how well they work for the user. But maybe I should wait with that a little longer, until we've got some more active users.

If anything of the above catches your eye, feel free to contribute! :)

**Reply by totti4ever** [reply]
regarding the survey: thought about adding optional and opt-in telemetry? I don't mind enabling this in well-documented open-source software. Firefly III does this (https://docs.firefly-iii.org/support/telemetry) and also show which data is sent exactly. Could give you some detailed insights in usage (amount and types of view, number of documents in total/per view, number of tags/correspondants, type of matching algorithm used, ...

But yeah, first and easier step would be a survey :-)

**Reply by jonaswinkler** [reply]
No telemetry in paperless.

**Reply by Boldie** [reply]
> Well, off the top of my head, these are the main concerns I have with the project right now:
> 
>     * The most pressing issue, by far, is the fact that the entire front end is untested (apart from me clicking through all the pages occasionally). But since that is still subject to change, I haven't done anything in that direction. I don't even know how front end testing works.
>  ....

I have worked a little bit using selenium and python for such things (although I am not a python / JS expert, since I work in C++ normally). With selenium the frontend is running in the browser and the bindings in the background to python can be used to interact or get feedback or status stuff from the DOM Model. I have used it in the past to automate a kind of robot for querying stuff from javascript webpages. Originally selenium was developed for testing webpages and frontends.

Googling a short time around for a short introduction to show how it works briefly: 
https://selenium-python.readthedocs.io/getting-started.html

**Reply by Boldie** [reply]
Hey and really thumbs up: I think it is a good practice to think of testing more than implementing more and more stuff :)

**Reply by jonaswinkler** [reply]
> Hey and really thumbs up: I think it is a good practice to think of testing more than implementing more and more stuff :)

Thank you! I've invested some major time to get the entire back end tested. I think that's far more important than the front end, as far as data integrity goes.

I'm aware of selenium, but I just don't have the experience or time to do it.

**Comment by jonaswinkler** [comment]
Oh, and something I've forgot yesterday:

I'd like to get this project translated into multiple languages eventually, which involves:
- Identifying strings for translation in the front end files
- Generate translation files
- Have these files translated
- Have Django serve different versions of the front end based on what locale the user requested

I haven't done anything in that direction yet, apart from making sure that it's possible to serve localized versions of the front end with django.
